heron/statefulstorages/src/java/com/twitter/heron/statefulstorage/hdfs/HDFSStorage.java

```diff
@@ -0,0 +1,188 @@
+// limitations under the License.
+
+package com.twitter.heron.statefulstorage.hdfs;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import com.twitter.heron.common.basics.SysUtils;
+import com.twitter.heron.proto.ckptmgr.CheckpointManager;
+import com.twitter.heron.spi.statefulstorage.Checkpoint;
+import com.twitter.heron.spi.statefulstorage.IStatefulStorage;
+import com.twitter.heron.spi.statefulstorage.StatefulStorageException;
+
+/**
+ * Note: The hadoop cluster config should be provided through the classpath
+ */
+public class HDFSStorage implements IStatefulStorage {
+  private static final Logger LOG = Logger.getLogger(HDFSStorage.class.getName());
+
+  private static final String ROOT_PATH_KEY = "heron.stateful.storage.hdfs.root.path";
+
+  private String checkpointRootPath;
+  private FileSystem fileSystem;
+
+  @Override
+  public void init(Map<String, Object> conf) throws StatefulStorageException {
+    LOG.info("Initializing... Config: " + conf.toString());
+    LOG.info("Class path: " + System.getProperty("java.class.path"));
+
+    checkpointRootPath = (String) conf.get(ROOT_PATH_KEY);
+
+    // Notice, we pass the config folder via classpath
+    // So hadoop will automatically search config files from classpath
+    Configuration hadoopConfig = new Configuration();
+
+    try {
+      fileSystem = FileSystem.get(hadoopConfig);
+      LOG.info("Hadoop FileSystem URI: " + fileSystem.getUri()
+          + " ; Home Dir: " + fileSystem.getHomeDirectory());
+    } catch (IOException e) {
+      throw new StatefulStorageException("Failed to get hadoop file system", e);
+    }
+  }
+
+  @Override
+  public void close() {
+    SysUtils.closeIgnoringExceptions(fileSystem);
+  }
+
+  @Override
+  public void store(Checkpoint checkpoint) throws StatefulStorageException {
+    Path path = new Path(getCheckpointPath(checkpoint));
+
+    // We need to ensure the existence of directories structure,
+    // since it is not guaranteed that FileSystem.create(..) always creates parents' dirs.
+    String checkpointDir = getCheckpointDir(checkpoint);
+    if (!createDirs(checkpointDir)) {
+      throw new StatefulStorageException("Failed to create dir: " + checkpointDir);
+    }
+
+    FSDataOutputStream out = null;
+    try {
+      out = fileSystem.create(path);
+      checkpoint.checkpoint().writeTo(out);
+    } catch (IOException e) {
+      throw new StatefulStorageException("Failed to persist", e);
+    } finally {
+      SysUtils.closeIgnoringExceptions(out);
+    }
+  }
+
+  @Override
+  public void restore(Checkpoint checkpoint) throws StatefulStorageException {
```

Comment:
Can we change this method so it doesn't mutate the state of the checkpoint passed to it and instead have it return a new Checkpoint? Generally it's preferred to favor immutability and to have methods without side effects. It also makes the contract very clear to the caller that it is getting a new checkpoint in return that is different from the one passed.